lib/AST/Attr.cpp

```diff
@@ -370,6 +370,17 @@ DeclAttribute *DeclAttribute::clone(ASTContext &ctx) const {
   }
 }
 
+bool DeclAttribute::canClone() const {
+  switch (getKind()) {
+#define DECL_ATTR(_,CLASS, ...)                                \
+  case DeclAttrKind::CLASS:                                    \
+    if (&CLASS##Attr::canClone == &DeclAttribute::canClone)    \
+      return true;                                             \
+    return static_cast<const CLASS##Attr *>(this)->canClone();
```

Comment:
Out of curiosity, why this manual dynamic dispatch rather than a virtual function? Does `DeclAttribute` not have a vtable or something?

Reply:
That's right: there's no vtable, so I don't want to add one.